## Desafío 1

#### Grupo
 Juan Manuel, Rafael, Diego Cabrera
#### Objetivos:
-  Efectuar una limpieza del dataset provisto. Particularmente, deberá diseñar estrategias para lidiar con los datos perdidos en ciertas variables.
-  Realizar un análisis descriptivo de las principales variables.
-  Crear nuevas columnas a partir de las características dadas que puedan tener valor predictivo.

 



In [2]:
import numpy as np
import pandas as pd

#import sys
# sys.setdefaultencoding() does not exist, here!
#reload(sys)  # Reload does the trick!
#sys.setdefaultencoding('UTF8')

df = pd.read_csv('data/properatti.csv.gz', compression='gzip', header=0)


In [181]:
#Print columns
df

operation property_type             place country_name  \
0           sell            PH         mataderos    Argentina   
1           sell     apartment          la plata    Argentina   
2           sell     apartment         mataderos    Argentina   
3           sell            PH           liniers    Argentina   
4           sell     apartment     mar del plata    Argentina   
5           sell         house       gualeguaych    Argentina   
6           sell            PH      vicente lpez    Argentina   
7           sell     apartment          belgrano    Argentina   
8           sell     apartment          belgrano    Argentina   
9           sell         house           rosario    Argentina   
10          sell         house            crdoba    Argentina   
11          sell         house        san miguel    Argentina   
12          sell     apartment        san isidro    Argentina   
13          sell     apartment           palermo    Argentina   
14          sell     apartment           palermo    Argentina   
15          sell         house        san miguel    Argentina   
16          sell            PH         mataderos    Argentina   
17          sell     apartment             tigre    Argentina   
18          sell         house             tigre    Argentina   
19          sell     apartment           palermo    Argentina   
20          sell         house    villa giardino    Argentina   
21          sell     apartment           palermo    Argentina   
22          sell         house  general rodrguez    Argentina   
23          sell     apartment             pilar    Argentina   
24          sell     apartment              morn    Argentina   
25          sell     apartment              morn    Argentina   
26          sell         house             tigre    Argentina   
27          sell         house             pilar    Argentina   
28          sell     apartment              morn    Argentina   
29          sell     apartment            flores    Argentina   
...          ...           ...               ...          ...   
121190      sell     apartment            crdoba    Argentina   
121191      sell     apartment            crdoba    Argentina   
121192      sell     apartment            crdoba    Argentina   
121193      sell     apartment            crdoba    Argentina   
121194      sell     apartment            crdoba    Argentina   
121195      sell     apartment            crdoba    Argentina   
121196      sell     apartment            crdoba    Argentina   
121197      sell     apartment            crdoba    Argentina   
121198      sell     apartment            crdoba    Argentina   
121199      sell     apartment            crdoba    Argentina   
121200      sell     apartment            crdoba    Argentina   
121201      sell     apartment            crdoba    Argentina   
121202      sell     apartment            crdoba    Argentina   
121203      sell     apartment            crdoba    Argentina   
121204      sell     apartment            crdoba    Argentina   
121205      sell     apartment            crdoba    Argentina   
121206      sell     apartment            crdoba    Argentina   
121207      sell         house           escobar    Argentina   
121208      sell         house           escobar    Argentina   
121209      sell     apartment          belgrano    Argentina   
121210      sell         house        san isidro    Argentina   
121211      sell     apartment        san isidro    Argentina   
121212      sell         house        san isidro    Argentina   
121213      sell         house        san isidro    Argentina   
121214      sell         store        san isidro    Argentina   
121215      sell     apartment          belgrano    Argentina   
121216      sell         house        san isidro    Argentina   
121217      sell     apartment     villa urquiza    Argentina   
121218      sell     apartment     mar del plata    Argentina   
121219      sell     apartment   capital f

####  Limpieza del dataset
   - a. Listado de columnas a eliminar. Armar funcion
   - b. Si hiciera falta, renombrar columnas para que sean mas representativas 
   
####  Analisis descriptivo
   - a. 
####  Nuevas columnas y valores predictivos
   - a. Armar fnc para crear automaticamente columnas para usar de flag para saber si se tocan datos
   - b.

In [177]:
df['price'].describe()

count    1.166510e+05
mean     2.627985e+05
std      8.739233e+05
min      0.000000e+00
25%      9.000000e+04
50%      1.490000e+05
75%      2.750000e+05
max      1.418258e+08
Name: price, dtype: float64

In [180]:
df.count()

operation                121220
property_type            121220
place                    121220
country_name             121220
state_name               121220
geonames_id              102503
lat                       69670
lon                       69670
price                    116651
surface_total_in_m2       82298
surface_covered_in_m2    101264
price_per_m2             120209
floor                      7899
rooms                     92973
expenses                  14262
description              121218
price_flag               121220
rooms_flag               121220
place_flag               121220
dtype: int64

In [276]:
#FUNCIONES 

#1. Elimina columnas
def drop_columns(df, serie):
    #print ('Eliminando columna/s ' + serie )
    df = df.drop(columns=serie,axis=1)
    return df
#2. Crea columnas "flag"
def make_flags(df, cols):
    col = {c: [add_flag(df, str(c))]
            for c in cols}

def add_flag(df, column): 
       # Creamos el nombre de la columna FLAG
    new_col = column + '_flag'
    # Creamos la columna con los valor 1 si la columna original tiene el registro null; sino, valor 0
    df[new_col] = df[column].isnull().astype(np.int8)
    print ('Columna ' + new_col +' creada' )

#3. Renombra columna
def rename_columns(df, cols):
    df = df.rename(cols, axis='columns')
    return df
    
#4. Elimina caracteres non-ascii   
def _removeNonAscii(s): return "".join(i for i in s if ord(i)<128)

#5. Calcula promedio por lugar/zona
def get_mean (state, nplace):
    global cnt_success
    global cnt_error
    try:
        cnt_success+= 1;
        return df_price_per_m2.loc[state, nplace]['mean']
    except Exception:
        #print 'error'
        #print('Error en ' +str(nplace))
        cnt_error+= 1;
        pass
        return np.nan

#6. calcula_ambientes en base a los mts2
def calcula_ambientes_xmt2 (mts):
    try:
       val = int(mts)
    except ValueError:
        return np.nan; #print("That's not an int!. Value: " + str(mts))
    val;

    if val <40:
       return 1;
    elif val <60:
       return 2;
    elif  val <80:
       return 3;
    elif  val <100:
       return 4;
    elif  val <120:
       return 5;
    elif  val >120:
       return 6;
    else:
       return np.nan;
# 7. calcula_mts2_xambientes
def calcula_mts2_xambientes (mts):
    try:
       val = int(mts)
    except ValueError:
        return np.nan; #print("That's not an int!. Value: " + str(mts))
    val;

    if val == 1:
       return 40;
    elif val == 2:
       return 60;
    elif  val == 3:
       return 80;
    elif  val == 4:
       return 100;
    elif  val == 5:
       return 120;
    elif  val >5:
       return 150;
    else:
       return np.nan;

#8. Calcula numero de ambientes en base a string 
def uno():
    return 1
def un():
    return 1
def dos():
    return 2
def tres():
    return 3
def cuatro():
    return 4
def cinco():
    return 5
def seis():
    return 6
def siete():
    return 7

switcher = {
        'un': uno,
        'uno': uno,
        'tres': tres,
        'cuatro': cuatro,
        'cinco': cinco,
        'seis': seis,
        'siete': siete
    }

def strings_to_number(argument,subfnc):
     try:
        str(argument)
        if argument <> 'nan':
            print 'pricessing ' + argument
    # Get the function from switcher dictionary
        func = switcher.get(argument,  lambda: np.nan)
    # Execute the function
        numb = func()
        if (subfnc == 'dorm') & (numb.is_integer()):
            numb = numb + 1  #Le sumo 1: 2 dormitorios = 3 ambientes
        return numb
     except ValueError:
        return np.nan; #print("That's not an int!. Value: " + str(mts))
    

In [5]:
# Variables GLobals
#Columnas a eliminar:
    #price, price_aprox_local_currency y currency: Tomamos en cuenta solo el valor en dolares (columna price_aprox_usd)
    #price_per_m2: Tomamos en cuenta solo el valor de price_usd_per_m2
COLUMNAS_A_ELIMINAR=["Unnamed: 0","lat-lon","currency","price_per_m2","price_aprox_local_currency","price","properati_url","title","image_thumbnail"];
#Columnas en las que llenaremos datos faltantes. ENtonces, creamos  columna que usaremos de flag para registrar los datos que tocamos
COLUMNAS_FLAG=['price','rooms','place']
#COlumnas a renombrar
COLUMNAS_A_RENOMBRAR={'price_aprox_usd':'price','place_name':'place','price_usd_per_m2':'price_per_m2'}
#Utilizado en fnc get_mean
cnt_success = 0
cnt_error = 0

In [6]:
#Elimino columnas
df = drop_columns(df, COLUMNAS_A_ELIMINAR)
#Renombro columnas
df = rename_columns(df, COLUMNAS_A_RENOMBRAR)

In [7]:
#Creo nuevas columnas "flag"
make_flags(df,COLUMNAS_FLAG)

Columna price_flag creada
Columna rooms_flag creada
Columna place_flag creada


In [293]:
#Valores nulos por columna
df.isnull().sum().sort_values()

operation                     0
price_flag                    0
rooms_flag                    0
state_name                    0
place_flag                    0
place                         0
property_type                 0
country_name                  0
description                   2
price_per_m2               1011
price                      1494
rooms                      3537
surface_covered_in_m2     16881
geonames_id               18717
surface_total_in_m2       38922
lat                       51550
lon                       51550
expenses                 106958
floor                    113321
dtype: int64

Total registros: 121220

Valores necesarios: 

- place_name 

Barrio/3 nivel de ubicacion; 1er y 2do nivel está completo (country_name/state_name). Tenemos completo -place_with_parent_names, de ahi se puede extraer el 3er nivel
- surface_total_in_m2
- price
- price_per_m2
- rooms

# PLACE NAME

In [194]:
df["place"].isnull().value_counts()

False    121220
Name: place, dtype: int64

In [10]:
#Verifico la 3 columna de place_with_parent_names
df[df["place"].isnull()]["place_with_parent_names"].str.split('|').str[3].value_counts()

Tigre    23
Name: place_with_parent_names, dtype: int64

In [11]:
#Cargo los datos de place_name y elimino la columna place_with_parent_names
df["place"] = df["place_with_parent_names"].apply(lambda place_with_parent_names: place_with_parent_names.split('|')[3] if (place_with_parent_names.split('|')[3]) else (place_with_parent_names.split('|')[2]))
#Elimino place_with_parent_names
df = drop_columns(df, "place_with_parent_names")

In [12]:
df['place'].value_counts().head(10)

Córdoba            9254
Tigre              8983
Rosario            8504
Mar del Plata      7710
Palermo            4083
Vicente López      3663
Lomas de Zamora    3470
La Plata           3360
Pilar              3252
San Isidro         3009
Name: place, dtype: int64

In [280]:
#Valores nulos por columna
df.isnull().sum().sort_values()

operation                     0
price_flag                    0
rooms_flag                    0
state_name                    0
place_flag                    0
place                         0
property_type                 0
country_name                  0
description                   2
price_per_m2               1011
price                      1698
rooms                      3537
surface_covered_in_m2     16881
geonames_id               18717
surface_total_in_m2       38922
lat                       51550
lon                       51550
expenses                 106958
floor                    113321
dtype: int64

# Rooms

Para los casos donde no tenemos el valor "rooms", intentaremos sacar informacion de la columna descripcion. Creamos una nueva columna como flag para indicar si rooms no estaba cargado

In [15]:
regexp="(\d+)(\s|())amb"
#a. Rooms null y en description aparece N amb*, pongo N como valor en rooms
df.loc[df['rooms'].isnull(),'rooms'] = df.loc[(df["rooms"].isnull()),"description"].str.lower().str.extract(regexp)[0]

#b. en la descripcion aparece "monoambiente" o "mono" y rooms es null, pongo 1
df.loc[((df['rooms'].isnull()) & (df['description'].str.lower().str.contains('monoambiente')) | df['description'].str.lower().str.contains('mono')),'rooms'] = 1

#c. x Dormitorios
regexp="(\d+)(\s|())dorm"
df.loc[df['rooms'].isnull(),'rooms'] = df.loc[(df["rooms"].isnull()),"description"].str.lower().str.extract(regexp)[0]




In [216]:
df['description'][538]#1


'Monoambiente en VentaPiso 10 Contrafrente Orientaci\xc3\xb3n OesteBa\xc3\xb1o CompletoBalcon amplioENTREGA ABRIL 2017El edificio dispone de una superficie total a edificar de 2290 m2, distribuidos en planta baja y diez pisos. En planta baja se ubica el palier de ingreso a los departamentos y ocho cocheras individuales. Cada uno de los diez pisos de departamentos esta compuesto por cuatro unidades, dos unidades de un dormitorio y dos monoambientes.Este proyecto se encuentra ubicado a una cuadra de Av. Pellegrini, a cuatro de Av. Belgrano y a siete de Bv. 27 de Febrero con accesos rapidos al macro y microcentro de la ciudad, tanto como a la Av. Circunvalacion y sus accesos a las autopistas a Buenos Aires, Cordoba y Santa Fe.'

In [198]:
#Revisamos el valor "surface_covered_in_m2" para estimar los ambientes. 

df.loc[df["rooms"].isnull() & df["surface_covered_in_m2"].notnull(),"rooms"] =  \
df.loc[df["rooms"].isnull() & df["surface_covered_in_m2"].notnull(),"surface_covered_in_m2"].apply(lambda surface_covered_in_m2: calcula_ambientes_xmt2(surface_covered_in_m2))

#Sino, nos basamos en el valor "surface_total_in_m2" para estimar los ambientes. 

df.loc[df["rooms"].isnull() & df["surface_covered_in_m2"].isnull() & df["surface_total_in_m2"].notnull() ,"rooms"] =  \
df.loc[df["rooms"].isnull() & df["surface_covered_in_m2"].isnull() & df["surface_total_in_m2"].notnull() ,"surface_total_in_m2"] .apply(lambda surface_total_in_m2: calcula_ambientes_xmt2(surface_total_in_m2))


In [277]:
# Calculo cantidad de rooms en base al string. Ej. "dos ambientes"
# no especifica cantidad de dormitorios:
regexp="(\w+) amb"

#df.loc[df["rooms"].isnull(),"description"].str.lower().str.extract(regexp)[0]

df.loc[df["rooms"].isnull(),"rooms"] =  \
df.loc[df["rooms"].isnull(),"description"].str.lower().str.extract(regexp)[0].apply(lambda description: strings_to_number(str(description),'amb'))
#Mismo caso, pero esta vez en base a dormitorios. Ej. "un dormitorio" = 2 ambientes
regexp="(\w+) dorm"

#df.loc[df["rooms"].isnull(),"description"].str.lower().str.extract(regexp)[0]

df.loc[df["rooms"].isnull(),"rooms"] =  \
df.loc[df["rooms"].isnull(),"description"].str.lower().str.extract(regexp)[0].apply(lambda description: strings_to_number(str(description),'dorm'))


pricessing de
pricessing dos
pricessing dos
pricessing amplio
pricessing los
pricessing los
pricessing los
pricessing dos
pricessing los
pricessing dos
pricessing los
pricessing para
pricessing por
pricessing los
pricessing los
pricessing esta
pricessing sobre
pricessing sobre
pricessing ambiente
pricessing dos
pricessing dos
pricessing dos
pricessing dos
pricessing amplios
pricessing dos
pricessing que
pricessing los
pricessing los
pricessing los
pricessing los
pricessing dos
pricessing dos
pricessing solo
pricessing los
pricessing los
pricessing dos
pricessing los
pricessing dos
pricessing en
pricessing los
pricessing dos
pricessing en
pricessing dos
pricessing dos
pricessing los
pricessing los
pricessing pulir
pricessing dos
pricessing el
pricessing dos
pricessing los
pricessing dos
pricessing dos
pricessing a
pricessing dos
pricessing dos
pricessing en
pricessing los
pricessing dos
pricessing dos
pricessing solo
pricessing amplio
pricessing dos
pricessing amplios
pricessing medio
p

pricessing 3m
pricessing placard
pricessing 50m
pricessing dos
pricessing y
pricessing dos
pricessing primer
pricessing alta
pricessing segundo
pricessing dos
pricessing dos
pricessing dos
pricessing dos
pricessing 5m
pricessing o
pricessing 2do
pricessing dos
pricessing dos
pricessing 50m
pricessing 50m
pricessing 50m
pricessing dos
pricessing dos
pricessing dos
pricessing en
pricessing en
pricessing dos
pricessing amplios
pricessing amplios
pricessing amplios
pricessing en
pricessing dos
pricessing dos
pricessing dos
pricessing dos
pricessing dos
pricessing dos
pricessing amplios
pricessing ambos
pricessing dos
pricessing dos
pricessing dos
pricessing dos
pricessing medio
pricessing medio
pricessing medio
pricessing dos
pricessing dos
pricessing dos
pricessing dos
pricessing dos
pricessing dos
pricessing dos
pricessing dos
pricessing dos
pricessing dos
pricessing dos
pricessing dos
pricessing 2do
pricessing 3er
pricessing dos
pricessing dos
pricessing dos
pricessing segundo
pricessin

In [215]:
df.loc[538]

operation                                                             sell
property_type                                                    apartment
place                                                              rosario
country_name                                                     Argentina
state_name                                                        santa fe
geonames_id                                                    3.83857e+06
lat                                                                -32.961
lon                                                               -60.6273
price                                                                  NaN
surface_total_in_m2                                                    NaN
surface_covered_in_m2                                                  NaN
price_per_m2                                                       2339.03
floor                                                                  NaN
rooms                    

# Price

1. Price null pero tengo surface_in_m2 y price_per_m2 entonces calculo precio
2. Calculo precio promedio por state_name/place
3. Si no tengo price ni price_per_m2 pero si surface_in_m2, entonces calculo price en base al promedio por place


In [287]:
#Price null pero tengo surface_in_m2 y price_per_m2 entonces calculo precio
df.loc[df['price'].isnull() & df["surface_total_in_m2"].notnull() & df["surface_total_in_m2"].notnull(),"price"] =  \
df.loc[df['price'].isnull() & df["surface_total_in_m2"].notnull() & df["price_per_m2"].notnull()].apply(lambda row: row.surface_total_in_m2 * row.price_per_m2, axis=1)

In [184]:
df.loc[df['price'].isnull()]

operation property_type                    place country_name  \
5           sell         house              gualeguaych    Argentina   
125         sell     apartment   bs.as. g.b.a. zona sur    Argentina   
126         sell         house               la matanza    Argentina   
133         sell     apartment                    salta    Argentina   
134         sell     apartment                    salta    Argentina   
136         sell         house                 santa fe    Argentina   
176         sell     apartment             general roca    Argentina   
180         sell         house                   emilia    Argentina   
201         sell         house                 santa fe    Argentina   
273         sell     apartment                  quilmes    Argentina   
275         sell         store                  quilmes    Argentina   
295         sell         house                santo tom    Argentina   
337         sell         house  san carlos de bariloche    Argentina   
338         sell     apartment             general roca    Argentina   
339         sell     apartment             general roca    Argentina   
445         sell     apartment                    salta    Argentina   
467         sell         house                 santa fe    Argentina   
468         sell         house                santo tom    Argentina   
470         sell         house            mar del plata    Argentina   
482         sell         house                santo tom    Argentina   
484         sell         house                 la plata    Argentina   
486         sell     apartment                esperanza    Argentina   
487         sell         house                esperanza    Argentina   
526         sell     apartment                  rosario    Argentina   
527         sell         house                  rosario    Argentina   
528         sell     apartment                  rosario    Argentina   
529         sell     apartment                  rosario    Argentina   
530         sell     apartment          tres de febrero    Argentina   
531         sell     apartment          tres de febrero    Argentina   
532         sell     apartment          tres de febrero    Argentina   
...          ...           ...                      ...          ...   
120448      sell     apartment                   crdoba    Argentina   
120450      sell     apartment                   crdoba    Argentina   
120486      sell     apartment                   crdoba    Argentina   
120490      sell         house                   crdoba    Argentina   
120521      sell     apartment                   crdoba    Argentina   
120523      sell         store                   crdoba    Argentina   
120544      sell     apartment                   crdoba    Argentina   
120551      sell         house                   crdoba    Argentina   
120552      sell     apartment            villa allende    Argentina   
120560      sell         house               mendiolaza    Argentina   
120562      sell         house               mendiolaza    Argentina   
120569      sell         house               mendiolaza    Argentina   
120580      sell         house               mendiolaza    Argentina   
120583      sell     apartment            villa allende    Argentina   
120587      sell     apartment                   crdoba    Argentina   
120601      sell         store            villa allende    Argentina   
120628      sell     apartment                   crdoba    Argentina   
120634      sell         house                   crdoba    Argentina   
120639      sell         house            villa allende    Argentina   
120641      sell         house            villa allende    Argentina   
120653      sell     apartment                   crdoba    Argentina   
120655      sell         house                   crdoba    Argentina   
120667      sell     apartment                   crdoba    Argentina   
120668      sell     apartment                 

In [30]:
#Elimino caracteres non ascii de place
df.loc[df['place'].notnull(),'place'] = df.loc[df['place'].notnull(),'place'].apply(lambda place: _removeNonAscii(place.lower()))
#Elimino caracteres non ascii de state_name
df.loc[df['state_name'].notnull(),'state_name'] = df.loc[df['state_name'].notnull(),'state_name'].apply(lambda state_name: _removeNonAscii(state_name.lower()))


In [31]:
#Creo df con estadisticas por lugar
#df_price_per_m2 = df[df['price_per_m2'].notnull()].groupby('place').agg(['min', 'max','mean','size','std'])['price_per_m2']

df_price_per_m2 = df[df['price_per_m2'].notnull()].groupby(['state_name','place']).agg(['min', 'max','mean','size','std'])['price_per_m2']

In [32]:
df[df['price'].isnull()].groupby('place').size().sort_values(ascending=False).head(10)

place
rosario                    3769
crdoba                     2338
la plata                   1091
mar del plata               787
tigre                       741
ituzaing                    726
palermo                     535
morn                        499
belgrano                    458
san carlos de bariloche     356
dtype: int64

In [33]:
df_price_per_m2.head()

min           max  \
state_name               place                                                
bs.as. g.b.a. zona norte bs.as. g.b.a. zona norte   16.958095  11333.333333   
                         escobar                    18.599562  16428.571429   
                         general san martn         121.396055  48296.222222   
                         jos c paz                  79.083333  26315.789474   
                         malvinas argentinas        23.863436   2750.809061   

                                                          mean  size  \
state_name               place                                         
bs.as. g.b.a. zona norte bs.as. g.b.a. zona norte  1310.303764   222   
                         escobar                   1110.781351  1909   
                         general san martn         2064.219071  1582   
                         jos c paz                  841.071808   140   
                         malvinas argentinas       1028.003798   489   

                                                           std  
state_name               place                                  
bs.as. g.b.a. zona norte bs.as. g.b.a. zona norte  1096.804263  
                         escobar                    640.318976  
                         general san martn         4755.982727  
                         jos c paz                 2213.395567  
                         malvinas argentinas        502.453770

In [34]:

    
    #df_price_per_m2 = df[df['price_per_m2'].notnull() & df['place'] == str(nplace) ].groupby('place').agg(['mean'])['price_per_m2']
    #print df_price_per_m2
    #return df_price_per_m2
    #return df_price_per_m2.loc[str (nplace)]['mean']

df.loc[df["price_per_m2"].isnull(),'price_per_m2'] =  \
df.loc[df['price_per_m2'].isnull()].apply(lambda row: get_mean(row.state_name, row.place), axis=1)


print 'Successfully: '  + str(cnt_success)
print 'Errors: ' + str(cnt_error)
#df.loc[df['price_per_m2'].isnull()]['place']

Successfully: 54625
Errors: 3033


In [35]:
df.loc[df['price_per_m2'].isnull()]

operation property_type                      place country_name  \
5           sell         house                gualeguaych    Argentina   
180         sell         house                     emilia    Argentina   
486         sell     apartment                  esperanza    Argentina   
487         sell         house                  esperanza    Argentina   
497         sell     apartment                  rivadavia    Argentina   
498         sell     apartment                  rivadavia    Argentina   
499         sell     apartment                  rivadavia    Argentina   
679         sell     apartment                  rivadavia    Argentina   
915         sell         house               bella italia    Argentina   
1052        sell         house              san francisco    Argentina   
1418        sell         house              san francisco    Argentina   
1478        sell         house                  rivadavia    Argentina   
1479        sell     apartment                  rivadavia    Argentina   
1592        sell     apartment                castellanos    Argentina   
1593        sell     apartment                castellanos    Argentina   
1622        sell     apartment                  rivadavia    Argentina   
1623        sell     apartment                  rivadavia    Argentina   
1805        sell         house                las talitas    Argentina   
1921        sell         house            carmen de areco    Argentina   
1930        sell         house                 las tapias    Argentina   
1935        sell         house                  los pozos    Argentina   
2002        sell         house  santa rosa de calamuchita    Argentina   
2065        sell         house  santa rosa de calamuchita    Argentina   
2459        sell         house                   vistalba    Argentina   
2736        sell         house                    achiras    Argentina   
2816        sell     apartment                 roque prez    Argentina   
2826        sell         house  santa rosa de calamuchita    Argentina   
2836        sell     apartment                   san juan    Argentina   
2946        sell         house                     toledo    Argentina   
3023        sell         house  santa rosa de calamuchita    Argentina   
...          ...           ...                        ...          ...   
115975      sell         house                  sinsacate    Argentina   
115990      sell         house              valle hermoso    Argentina   
116011      sell     apartment              valle hermoso    Argentina   
116017      sell         house              villa fontana    Argentina   
116032      sell         house                  mi granja    Argentina   
116037      sell         house  santa rosa de calamuchita    Argentina   
116049      sell         house  santa rosa de calamuchita    Argentina   
116056      sell         house                 ro segundo    Argentina   
116062      sell         house         villa de las rosas    Argentina   
116108      sell         house           falda del carmen    Argentina   
116117      sell         house                  anisacate    Argentina   
116127      sell         house  santa rosa de calamuchita    Argentina   
116136      sell         house                salsipuedes    Argentina   
116326      sell         house               arroyo leyes    Argentina   
116327      sell         house                castellanos    Argentina   
116350      sell     apartment                  esperanza    Argentina   
116479      sell         house                      chaco    Argentina   
117108      sell         house               san clemente    Argentina   
117146      sell         house                   oncativo    Argentina   
117229      sell         house                general paz    Argentina   
119320      sell     apartment                castellanos    Argentina   
119321      sell     apartment                castellanos    Argentina   
119567     

In [43]:

df_price_per_m2.loc['santa fe',]['mean']

place
alvear                        403.350150
arequito                     1438.356164
arroyo seco                   264.550265
capitn bermdez                463.231674
coronda                        76.771599
empalme villa constitucin     455.679833
fighiera                      196.292850
franck                        280.418379
fray luis beltrn              352.244821
funes                         791.644734
glvez                         240.358600
granadero baigorria           591.623643
ibarlucea                     349.873118
iriondo                       747.533501
las rosas                     695.373443
los laureles                 1362.551020
malabrigo                     467.363967
mara juana                    190.684480
monje                          91.192959
monte vera                      9.388501
murphy                        468.750000
prez                          990.000000
providencia                  2216.216216
pueblo andino                  45.794393
rafaela   

In [ ]:
df[df['price_per_m2'].isnull()].value_counts()

In [59]:
df.loc[df['place'] == 'esperanza']

operation property_type      place country_name state_name  \
486         sell     apartment  esperanza    Argentina   santa fe   
487         sell         house  esperanza    Argentina   santa fe   
5155        sell     apartment  esperanza    Argentina   santa fe   
6378        sell            PH  esperanza    Argentina   santa fe   
8625        sell         house  esperanza    Argentina   santa fe   
8626        sell         house  esperanza    Argentina   santa fe   
9216        sell         house  esperanza    Argentina   santa fe   
10361       sell         store  esperanza    Argentina   santa fe   
10362       sell         house  esperanza    Argentina   santa fe   
10792       sell     apartment  esperanza    Argentina   santa fe   
10793       sell     apartment  esperanza    Argentina   santa fe   
12655       sell     apartment  esperanza    Argentina   santa fe   
12847       sell         house  esperanza    Argentina   santa fe   
12848       sell         house  esperanza    Argentina   santa fe   
13508       sell         house  esperanza    Argentina   santa fe   
15124       sell         house  esperanza    Argentina   santa fe   
15125       sell            PH  esperanza    Argentina   santa fe   
16308       sell         house  esperanza    Argentina   santa fe   
16309       sell     apartment  esperanza    Argentina   santa fe   
16310       sell     apartment  esperanza    Argentina   santa fe   
16672       sell     apartment  esperanza    Argentina   santa fe   
16673       sell     apartment  esperanza    Argentina   santa fe   
16674       sell     apartment  esperanza    Argentina   santa fe   
16675       sell     apartment  esperanza    Argentina   santa fe   
16676       sell     apartment  esperanza    Argentina   santa fe   
17072       sell     apartment  esperanza    Argentina   santa fe   
18838       sell     apartment  esperanza    Argentina   santa fe   
19126       sell     apartment  esperanza    Argentina   santa fe   
19127       sell     apartment  esperanza    Argentina   santa fe   
19418       sell         house  esperanza    Argentina   santa fe   
...          ...           ...        ...          ...        ...   
40803       sell         house  esperanza    Argentina   santa fe   
41218       sell         house  esperanza    Argentina   santa fe   
41219       sell         house  esperanza    Argentina   santa fe   
41220       sell         house  esperanza    Argentina   santa fe   
41221       sell         house  esperanza    Argentina   santa fe   
44451       sell         house  esperanza    Argentina   santa fe   
44452       sell         house  esperanza    Argentina   santa fe   
47424       sell         house  esperanza    Argentina   santa fe   
47425       sell         house  esperanza    Argentina   santa fe   
47934       sell     apartment  esperanza    Argentina   santa fe   
49042       sell         house  esperanza    Argentina   santa fe   
50973       sell     apartment  esperanza    Argentina   santa fe   
52246       sell         house  esperanza    Argentina   santa fe   
52922       sell         house  esperanza    Argentina   santa fe   
57044       sell         house  esperanza    Argentina   santa fe   
57046       sell         house  esperanza    Argentina   santa fe   
57050       sell     apartment  esperanza    Argentina   santa fe   
57763       sell            PH  esperanza    Argentina   santa fe   
57768       sell     apartment  esperanza    Argentina   santa fe   
58454       sell         house  esperanza    Argentina   santa fe   
58455       sell     apartment  esperanza    Argentina   santa fe   
58456       sell     apartment  esperanza    Argentina   santa fe   
59699       sell         house  esperanza    Argentina   santa fe   
59982       sell         house  esperanza    Argentina   santa fe   
61880       sell     apartment  esperanza    Argentina   santa fe   
61881       sell         house  esperanza    Argentina   santa fe   
63123

In [167]:
df.loc[df['price'].isnull()]

operation property_type                    place country_name  \
5           sell         house              gualeguaych    Argentina   
125         sell     apartment   bs.as. g.b.a. zona sur    Argentina   
126         sell         house               la matanza    Argentina   
133         sell     apartment                    salta    Argentina   
134         sell     apartment                    salta    Argentina   
136         sell         house                 santa fe    Argentina   
176         sell     apartment             general roca    Argentina   
180         sell         house                   emilia    Argentina   
201         sell         house                 santa fe    Argentina   
273         sell     apartment                  quilmes    Argentina   
275         sell         store                  quilmes    Argentina   
295         sell         house                santo tom    Argentina   
337         sell         house  san carlos de bariloche    Argentina   
338         sell     apartment             general roca    Argentina   
339         sell     apartment             general roca    Argentina   
445         sell     apartment                    salta    Argentina   
467         sell         house                 santa fe    Argentina   
468         sell         house                santo tom    Argentina   
470         sell         house            mar del plata    Argentina   
482         sell         house                santo tom    Argentina   
484         sell         house                 la plata    Argentina   
486         sell     apartment                esperanza    Argentina   
487         sell         house                esperanza    Argentina   
526         sell     apartment                  rosario    Argentina   
527         sell         house                  rosario    Argentina   
528         sell     apartment                  rosario    Argentina   
529         sell     apartment                  rosario    Argentina   
530         sell     apartment          tres de febrero    Argentina   
531         sell     apartment          tres de febrero    Argentina   
532         sell     apartment          tres de febrero    Argentina   
...          ...           ...                      ...          ...   
120448      sell     apartment                   crdoba    Argentina   
120450      sell     apartment                   crdoba    Argentina   
120486      sell     apartment                   crdoba    Argentina   
120490      sell         house                   crdoba    Argentina   
120521      sell     apartment                   crdoba    Argentina   
120523      sell         store                   crdoba    Argentina   
120544      sell     apartment                   crdoba    Argentina   
120551      sell         house                   crdoba    Argentina   
120552      sell     apartment            villa allende    Argentina   
120560      sell         house               mendiolaza    Argentina   
120562      sell         house               mendiolaza    Argentina   
120569      sell         house               mendiolaza    Argentina   
120580      sell         house               mendiolaza    Argentina   
120583      sell     apartment            villa allende    Argentina   
120587      sell     apartment                   crdoba    Argentina   
120601      sell         store            villa allende    Argentina   
120628      sell     apartment                   crdoba    Argentina   
120634      sell         house                   crdoba    Argentina   
120639      sell         house            villa allende    Argentina   
120641      sell         house            villa allende    Argentina   
120653      sell     apartment                   crdoba    Argentina   
120655      sell         house                   crdoba    Argentina   
120667      sell     apartment                   crdoba    Argentina   
120668      sell     apartment                 

In [288]:
#Calculo surface_covered_in_m2 y/o surface_total_in_m2 en base a la descripcion
regexp="(\d+)(\s|())mts(..)cub|(\d+)(\s|())mts(.)cub|(\d+)(\s|())metros cub"
#surface_covered_in_m2
df.loc[df['surface_covered_in_m2'].isnull() & df["description"].notnull() & df["price"].isnull(),'surface_covered_in_m2'] =  \
df.loc[df['surface_covered_in_m2'].isnull() & df["description"].notnull() & df["price"].isnull(),'description'].str.lower().str.extract(regexp)[0]
regexp="(\d+)(\s|())mts(..)cub|(\d+)(\s|())mts(.)cub|(\d+)(\s|())metros cub"
#surface_total_in_m2: surface_covered_in_m2 y surface_total_in_m2 Null
regexp="(\d+)(\s|())mts|(\d+)(\s|())metros"
df.loc[df['surface_total_in_m2'].isnull() & df['surface_covered_in_m2'].isnull() & df["description"].notnull() & df["price"].isnull(),'surface_total_in_m2'] =  \
df.loc[df['surface_total_in_m2'].isnull() & df['surface_covered_in_m2'].isnull() & df["description"].notnull() & df["price"].isnull(),'description'].str.lower().str.extract(regexp)[0]


In [53]:
#Valido por ejemplo esperanza santa fe, no tiene estadisticas
df.loc[df['place'] == 'esperanza'].groupby(['state_name','place']).agg(['min', 'max','mean','size','std'])['price_per_m2']

min  max  mean  size  std
state_name place                               
santa fe   esperanza  NaN  NaN   NaN    74  NaN

In [58]:
df.loc[df['price'].isnull() & df['price_per_m2'].notnull() & df['surface_total_in_m2'].notnull()]

operation property_type                   place country_name  \
20          sell         house          villa giardino    Argentina   
41          sell     apartment        villa carlos paz    Argentina   
62          sell         house             berazategui    Argentina   
67          sell         house                ituzaing    Argentina   
76          sell         house           mar del plata    Argentina   
83          sell     apartment                 rosario    Argentina   
84          sell     apartment                 rosario    Argentina   
85          sell     apartment                 rosario    Argentina   
100         sell     apartment                 rosario    Argentina   
101         sell     apartment                 rosario    Argentina   
102         sell     apartment                 rosario    Argentina   
103         sell         house                ituzaing    Argentina   
124         sell         house             san lorenzo    Argentina   
139         sell         house           mar del plata    Argentina   
184         sell     apartment                 rosario    Argentina   
185         sell     apartment                 rosario    Argentina   
186         sell     apartment                 rosario    Argentina   
197         sell     apartment                   lobos    Argentina   
199         sell            PH  san martn de los andes    Argentina   
214         sell     apartment               caballito    Argentina   
217         sell     apartment               caballito    Argentina   
223         sell     apartment                belgrano    Argentina   
224         sell            PH                 escobar    Argentina   
236         sell     apartment                 rosario    Argentina   
240         sell     apartment                 rosario    Argentina   
243         sell     apartment                 rosario    Argentina   
245         sell     apartment                 rosario    Argentina   
247         sell     apartment                 rosario    Argentina   
248         sell     apartment                 rosario    Argentina   
249         sell     apartment                 rosario    Argentina   
...          ...           ...                     ...          ...   
121038      sell     apartment                belgrano    Argentina   
121039      sell     apartment                belgrano    Argentina   
121040      sell     apartment                belgrano    Argentina   
121041      sell     apartment                belgrano    Argentina   
121042      sell     apartment                belgrano    Argentina   
121043      sell     apartment                belgrano    Argentina   
121044      sell     apartment                belgrano    Argentina   
121045      sell     apartment                belgrano    Argentina   
121046      sell     apartment                belgrano    Argentina   
121047      sell     apartment                belgrano    Argentina   
121048      sell     apartment                belgrano    Argentina   
121049      sell     apartment                belgrano    Argentina   
121050      sell     apartment                belgrano    Argentina   
121051      sell     apartment                belgrano    Argentina   
121052      sell     apartment                belgrano    Argentina   
121085      sell     apartment                 rosario    Argentina   
121087      sell         house             yerba buena    Argentina   
121099      sell     apartment                    morn    Argentina   
121110      sell     apartment                 rosario    Argentina   
121114      sell     apartment                    morn    Argentina   
121115      sell     apartment                    morn    Argentina   
121116      sell     apartment                    morn    Argentina   
121123      sell     apartment                    morn    Argentina   
121126      sell     apartment                  crdoba    Argentina   
121127      sell     apartment 

In [289]:
#Tengo price_per_m2 y surface_total_in_m2 pero no precio entonces calculo precio
df.loc[df['price'].isnull() & df['price_per_m2'].notnull() & df['surface_total_in_m2'].notnull(), 'price'] =  \
df.loc[df['price'].isnull() & df['price_per_m2'].notnull() & df['surface_total_in_m2'].notnull()].apply(lambda row: (np.int(row.price_per_m2) * np.int(row.surface_total_in_m2)), axis=1)

In [290]:
#Tengo price_per_m2 y surface_covered_in_m2 pero no precio entonces calculo precio
df.loc[df['price'].isnull() & df['price_per_m2'].notnull() & df['surface_covered_in_m2'].notnull(), 'price'] = \
df.loc[df['price'].isnull() & df['price_per_m2'].notnull() & df['surface_covered_in_m2'].notnull()].apply(lambda row: (np.int(row.price_per_m2) * np.int(row.surface_covered_in_m2)), axis=1)

In [291]:
#Cantidad de precios calculados hasta el momento
print 'Price:'
df[df['price_flag'] == 1]['price_flag'].sum() - df['price'].isnull().sum()


Price:


18916

In [292]:
print 'Rooms:'
df[df['rooms_flag'] == 1]['rooms_flag'].sum() - df['rooms'].isnull().sum()

Rooms:


70293

In [294]:
df.loc[(df['price'].isnull()) & (df['place'] == 'rosario')].sort_values(['state_name','place'],ascending=False)

operation property_type    place country_name state_name  geonames_id  \
526         sell     apartment  rosario    Argentina   santa fe    3838574.0   
527         sell         house  rosario    Argentina   santa fe    3838574.0   
528         sell     apartment  rosario    Argentina   santa fe    3838574.0   
4726        sell         store  rosario    Argentina   santa fe    3838574.0   
5001        sell     apartment  rosario    Argentina   santa fe    3838574.0   
7346        sell     apartment  rosario    Argentina   santa fe    3838574.0   
7347        sell     apartment  rosario    Argentina   santa fe    3838574.0   
7348        sell     apartment  rosario    Argentina   santa fe    3838574.0   
7349        sell     apartment  rosario    Argentina   santa fe    3838574.0   
7350        sell     apartment  rosario    Argentina   santa fe    3838574.0   
7351        sell     apartment  rosario    Argentina   santa fe    3838574.0   
7352        sell     apartment  rosario    Argentina   santa fe    3838574.0   
7353        sell     apartment  rosario    Argentina   santa fe    3838574.0   
7354        sell     apartment  rosario    Argentina   santa fe    3838574.0   
7355        sell     apartment  rosario    Argentina   santa fe    3838574.0   
7553        sell     apartment  rosario    Argentina   santa fe    3838574.0   
7554        sell     apartment  rosario    Argentina   santa fe    3838574.0   
7634        sell     apartment  rosario    Argentina   santa fe    3838574.0   
8066        sell         store  rosario    Argentina   santa fe    3838574.0   
8067        sell         house  rosario    Argentina   santa fe    3838574.0   
8068        sell     apartment  rosario    Argentina   santa fe    3838574.0   
8069        sell     apartment  rosario    Argentina   santa fe    3838574.0   
8884        sell     apartment  rosario    Argentina   santa fe    3838574.0   
9455        sell     apartment  rosario    Argentina   santa fe    3838574.0   
9457        sell     apartment  rosario    Argentina   santa fe    3838574.0   
10399       sell     apartment  rosario    Argentina   santa fe    3838574.0   
10400       sell     apartment  rosario    Argentina   santa fe    3838574.0   
10402       sell     apartment  rosario    Argentina   santa fe    3838574.0   
10404       sell     apartment  rosario    Argentina   santa fe    3838574.0   
10405       sell     apartment  rosario    Argentina   santa fe    3838574.0   
...          ...           ...      ...          ...        ...          ...   
110011      sell         house  rosario    Argentina   santa fe    3838574.0   
110016      sell     apartment  rosario    Argentina   santa fe    3838574.0   
110017      sell     apartment  rosario    Argentina   santa fe    3838574.0   
110020      sell         house  rosario    Argentina   santa fe    3838574.0   
110038      sell     apartment  rosario    Argentina   santa fe    3838574.0   
110043      sell         store  rosario    Argentina   santa fe    3838574.0   
110044      sell         store  rosario    Argentina   santa fe    3838574.0   
110048      sell         house  rosario    Argentina   santa fe    3838574.0   
110049      sell     apartment  rosario    Argentina   santa fe    3838574.0   
110053      sell         house  rosario    Argentina   santa fe    3838574.0   
110054      sell     apartment  rosario    Argentina   santa fe    3838574.0   
110879      sell         store  rosario    Argentina   santa fe    3838574.0   
110880      sell     apartment  rosario    Argentina   santa fe    3838574.0   
110881      sell     apartment  rosario    Argentina   santa fe    3838574.0   
111547      sell         store  rosario    Argentina   santa fe    3838574.0   
113221      sell         store  rosario    Argentina   santa fe    3838574.0   
113225      sell     apartment  rosario    Argentina   santa fe    3838574.0   
113755      sell     apartment  rosario    Argentina   santa fe    3838574.0   
114196 

In [279]:
#Cantidad de precios faltantes por lugar
df[df['price'].isnull()].groupby('place').size().sort_values(ascending=False).head(15)

place
rosario                      245
crdoba                       177
castellanos                  131
escobar                      105
la plata                      81
esperanza                     69
santa rosa de calamuchita     35
la matanza                    25
tres de febrero               22
santa fe                      21
san carlos de bariloche       20
coronda                       20
san francisco                 19
esteban echeverra             19
confluencia                   17
dtype: int64

In [286]:
#tengo ambiente y price_per_m2 y no tengo precio ni surface_covered_in_m2
#entonces estimo surface_covered_in_m2 en base a los ambientes y luego calculo precio

#df[df['price_per_m2'].notnull() & df['rooms'].notnull() & df['price'].isnull() & df['surface_covered_in_m2'].isnull()]['price_per_m2']

df.loc[df["price_per_m2"].notnull() & df["rooms"].notnull() & df['price'].isnull() & df['surface_covered_in_m2'].isnull(),"surface_covered_in_m2"] =  \
df.loc[df["price_per_m2"].notnull() & df["rooms"].notnull() & df['price'].isnull() & df['surface_covered_in_m2'].isnull(),"rooms"].apply(lambda rooms: calcula_mts2_xambientes(rooms))


In [231]:
#Vuelvo a ejecutar 'Tengo price_per_m2 y surface_covered_in_m2 pero no precio entonces calculo precio'

Series([], Name: price_per_m2, dtype: float64)

6